File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int start = 0;
        int end = nums.size() - 1;

        while(start <= end){
            int mid = (start + end)/2;
            if(nums[mid] == target){
                return true;
            }
            if(nums[mid] == nums[end] && mid!=end){
                end--;
                continue;
            }
            if(nums[start] == nums[mid] && start!=mid){
                start++;
                continue;
            }
            
            if(nums[mid] < nums[end]){
                if(target > nums[mid] && target <= nums[end]){
                    start = mid + 1;
                }
                else{
                    end = mid - 1;
                }
            }

            else{
                if(target >= nums[start] && target < nums[mid]){
                    end = mid - 1;
                }
                else{
                    start = mid + 1;
                }
            }
        }
        
        return false;
    }
};

File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.java
class Solution {
    public boolean search(int[] nums, int k) {
        
        int start = 0 ;
        int end = nums.length -1 ;

        while(start <= end){

            int mid = start + (end - start)/2 ;

            if(nums[mid] == k){
                return true;
            }

            if (nums[start] == nums[mid] && nums[mid] == nums[end]) {
                start++;
                end--;
                continue ;
            }


            if(nums[mid] >= nums[start]) // left side sorted hai
            {
                if(k >= nums[start] && k < nums[mid]){
                    end = mid - 1;
                }
                else{
                    start = mid + 1 ;
                }
            }


            else{ // right part sorted hai isme check kaar lo element ko


                if(k > nums[mid] && k <= nums[end]){
                    start  = mid + 1 ;
                }else{
                    end = mid - 1 ;
                }

            }
        }

        return false;

    }
}